File: src/main/scala/com/vc/hard/L980.scala
package com.vc.hard

object L980 {
  def uniquePathsIII(arr: Array[Array[Int]]): Int = {
    val dirs = Array((1, 0), (-1, 0), (0, 1), (0, -1))
    var res = 0
    var xs = 0
    var ys = 0
    var xe = 0
    var ye = 0
    var empty = 1
    val n = arr.length
    if(n == 0) return 0
    val m = arr(0).length
    def solve(x: Int, y: Int): Unit = {
      if(x == xe && y == ye && empty == 0) res += 1
      else {
        dirs.foreach(dir => {
          val xNew = dir._1 + x
          val yNew = dir._2 + y
          if(xNew >= 0 && xNew < n && yNew >= 0 && yNew < m && arr(xNew)(yNew) >= 0) {
            arr(xNew)(yNew) = -2
            empty -= 1
            solve(xNew, yNew)
            arr(xNew)(yNew) = 0
            empty += 1
          }
        })
      }
    }
    arr.indices.foreach(i => {
      arr(i).indices.foreach(j => {
        if(arr(i)(j) == 1) {
          xs = i
          ys = j
        }
        else if(arr(i)(j) == 2) {
          xe = i
          ye = j
        }
        else if(arr(i)(j) == 0) empty += 1
      })
    })
    arr(xs)(ys) = -2
    solve(xs, ys)
    res
  }
}

File: src/main/java/com/vc/hard/L980.java
package com.vc.hard;

class L980 {
    int count = 0;
    public int uniquePathsIII(int[][] grid) {
        int n = grid.length;
        if(n == 0) return 0;
        int m = grid[0].length;
        int empty = 1, sx = 0, sy = 0, ex = 0, ey = 0;
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if(grid[i][j] == 0) empty++;
                else if(grid[i][j] == 1) {
                    sx = i;
                    sy = j;
                }
                else if(grid[i][j] == 2) {
                    ex = i;
                    ey = j;
                }
            }
        }
        solve(sx, sy, ex, ey, empty, grid);
        return count;
    }

    private int[][] dirs = {{1, 0},{0, 1},{-1, 0},{0, -1}};
    private void solve(int sx, int sy, int ex, int ey, int empty, int[][] grid) {
        if(grid[sx][sy] == -2) return;
        if(sx == ex && sy == ey && empty == 0) count++;

        grid[sx][sy] = -2;
        for(int[] dir: dirs) {
            int xNew = sx + dir[0];
            int yNew = sy + dir[1];
            if(xNew >= 0 && xNew < grid.length && yNew >= 0 && yNew < grid[0].length
                    && grid[xNew][yNew] >= 0) {
                solve(xNew, yNew, ex, ey, empty - 1, grid);
            }
        }
        grid[sx][sy] = 0;
    }
}
